### PR TITLE
AIRO-1692 AIRO-1706 Retuning correct axis information

### PR DESCRIPTION
**Pull request by vidurvij-Unity**
## Proposed change(s)

Correcting axis export bug fix

### Useful links (GitHub issues, JIRA tickets, forum threads, etc.)

https://github.com/Unity-Technologies/URDF-Importer/issues/170
https://github.com/Unity-Technologies/URDF-Importer/issues/185

### Types of change(s)

- [x] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Documentation update
- [ ] Other (please describe)

## Testing and Verification

Please describe the tests that you ran to verify your changes. Please also provide instructions, ROS packages, and Unity project files as appropriate so we can reproduce the test environment. 

### Test Configuration:
- Unity Version: [e.g. Unity 2020.2.0f1]
- Unity machine OS + version: [e.g. Windows 10]
- ROS machine OS + version: [e.g. Ubuntu 18.04, ROS Noetic]
- ROS–Unity communication: [e.g. Docker]

## Checklist
- [x] Ensured this PR is up-to-date with the `dev` branch
- [x] Created this PR to target the `dev` branch
- [ ] Followed the style guidelines as described in the [Contribution Guidelines](https://github.com/Unity-Technologies/URDF-Importer/blob/main/CONTRIBUTING.md)
- [x] Added tests that prove my fix is effective or that my feature works
- [ ] Increased the [test coverage criteria](https://github.com/Unity-Technologies/URDF-Importer/blob/dev/.yamato/yamato-config.yml#L18) by 3%
- [ ] Updated the [Changelog](https://github.com/Unity-Technologies/URDF-Importer/blob/dev/com.unity.robotics.urdf-importer/CHANGELOG.md) and described changes in the [Unreleased section](https://github.com/Unity-Technologies/URDF-Importer/blob/dev/com.unity.robotics.urdf-importer/CHANGELOG.md#unreleased)
- [ ] Updated the documentation as appropriate

## Other comments

### CODE REVIEW

File: com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJointContinuous.cs [hunk 1]
```diff
@@ -103,7 +103,8 @@ protected override void ImportJointData(Joint joint)
         protected override Joint ExportSpecificJointData(Joint joint)
         {
 #if UNITY_2020_1_OR_NEWER
-            joint.axis = GetAxisData(axisofMotion);
+            Debug.Log($"{joint.parent} {axisofMotion} {joint.child}");
```
Comment (albertJkt) [inline com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJointContinuous.cs]: Debug.Log here looks like a leftover, or is it left intentionally?

File: com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJointContinuous.cs [hunk 2]
```diff
@@ -124,6 +132,7 @@ protected override Joint ExportSpecificJointData(Joint joint)
         protected override void AdjustMovement(Joint joint)
         {
             axisofMotion = joint.axis.xyz.ToVector3();
+            Debug.Log($"{joint.parent} {axisofMotion} {joint.child}");
```
Comment (albertJkt) [inline com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJointContinuous.cs]: 1 more Debug.Log here

File: com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJointContinuous.cs [hunk 3]
```diff
@@ -115,6 +116,13 @@ protected override Joint ExportSpecificJointData(Joint joint)
 #endif
             return joint;
         }
+        
+        protected Joint.Axis GetAxisData()
+        {
+            var res = (-1 * (unityJoint.anchorRotation * new Vector3(1, 0, 0)));//.Unity2Ros();
```
Comment (albertJkt) [inline com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJointContinuous.cs]: Unity2Ros() - looks like it needs to be uncommented to convert to LHS that's used in Unity.

**Review by albertJkt — COMMENTED** [review]
Left some comments. Mostly I'm concerned about commented Unity2Ros() call.

File: com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJointContinuous.cs [hunk 4]
```diff
@@ -115,15 +116,22 @@ protected override Joint ExportSpecificJointData(Joint joint)
 #endif
             return joint;
         }
+        
+        public override Joint.Axis GetAxisData()
+        {
+            var res = (-1 * (unityJoint.anchorRotation * new Vector3(1, 0, 0))).Unity2Ros();
```
Comment (K-Tone) [inline com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJointContinuous.cs]: res = unityJoint.anchorRotation * Vector3.left ?

File: com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJointPlanar.cs [hunk 5]
```diff
@@ -166,16 +172,38 @@ protected override bool IsJointAxisDefined()
...
+                    motionAxis = i;
+                    break;
+                }
             }
-            axisofMotion = new Vector3((float)joint.axis.xyz[0], (float)joint.axis.xyz[1], (float)joint.axis.xyz[2]);
-            int motionAxis = joint.axis.AxisofMotion();
+            
             Quaternion motion = unityJoint.anchorRotation;
 
+            switch (motionAxis)
```
Comment (K-Tone) [inline com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJointPlanar.cs:190]: is there a comment somewhere that states what 0, 1, 2 correspond to? x, y, z? any other order?

File: com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJoint.cs [hunk 6]
```diff
@@ -292,12 +295,6 @@ public void GenerateUniqueJointName()
             jointName = transform.parent.name + "_" + transform.name + "_joint";
         }
 
-        protected static Joint.Axis GetAxisData(Vector3 axis)
-        {
-            double[] rosAxis = axis.ToRoundedDoubleArray();
-            return new Joint.Axis(rosAxis);
-        }
```
Comment (K-Tone) [inline com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJoint.cs:299]: Why wouldn't we keep this in base class still, but specialise with unity joint? Seems all that the derived classes do is to use the same implementation basically (with the exception of one)

**Comment by albertJkt** [comment]
I've tested these changes locally using our Forklift project - axis information is exported and then reimported properly there. 

1 concern that is left regarding this PR (was also mentioned by Anthony) is why GetAxisData() was removed from the base class?

File: com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJoint.cs (hunk 6 above)
Comment (vidurvij-Unity) [reply com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJoint.cs:299]: Do you want me to use virtual override for that single case ?

File: com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJointPlanar.cs (hunk 5 above)
Comment (vidurvij-Unity) [reply com.unity.robotics.urdf-importer/Runtime/UrdfComponents/UrdfJoints/UrdfJointPlanar.cs:190]: Added it.

**Review by K-Tone — APPROVED** [review]

**Pull request merged**